In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [4]:
#reloading our prepared data

X_train_bal = pd.read_csv("../data/X_train_bal.csv")
Y_train_bal = pd.read_csv("../data/Y_train_bal.csv")
X_test = pd.read_csv("../data/X_test.csv")
Y_test = pd.read_csv("../data/Y_test.csv")

what all the metrics we test will mean?

1. Precision: of all the customers predicted as churners, how many of them were actually churners
2. Recall: of all the actual churners, how many did the model correctly identify
3. F1-score: The balance between precision and recall
4. Support: number of samples for that class in our test dataset

In [5]:
# Logistic Regression

log_reg = LogisticRegression(max_iter = 1000) #creating the model
log_reg.fit(X_train_bal, Y_train_bal) #training the model (.fit() means learn from the dataset)
y_pred_lr = log_reg.predict(X_test) #make predictions
print("Logistic Regression Accuracy: ", accuracy_score(Y_test, y_pred_lr)) #evaluate the model's performance
print(classification_report(Y_test, y_pred_lr)) #metrics report

C:\Users\KARISHMA\Desktop\customer-churn-prediction\venv\Lib\site-packages\sklearn\utils\validation.py:1406: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Logistic Regression Accuracy:  0.5387864161351491
              precision    recall  f1-score   support

           0       0.62      0.34      0.44      6168
           1       0.51      0.77      0.61      5434

    accuracy                           0.54     11602
   macro avg       0.56      0.55      0.52     11602
weighted avg       0.57      0.54      0.52     11602



what the above output means??

with the help of logistic regression model, we get 53.87% accuracy only. 
This means our model isn't performing that well.

from the above table we can see that the model is able to identify churners (recall = 0.77) more 
accurately than identifying non-churners.

what's the use of macro avg and weighted avg?
macro average: treats both classes equally (this is useful when the classes are balanced)
weighted average: weights by number of samples. (useful when one class dominates)

from the above result we can come to a conclusion that Logistic Regression might be too simple for our data, since our data might be complex or huge. So we need to test our data on other models.

In [6]:
# Random Forest

rf = RandomForestClassifier(n_estimators=200, random_state=42) #creating the model
rf.fit(X_train_bal, Y_train_bal) #training the model
y_pred_rf = rf.predict(X_test) #making predictions
print("Random Forest Accuracy: ", accuracy_score(Y_test, y_pred_rf)) #evaluating model's performance
print(classification_report(Y_test, y_pred_rf)) #metrics report

C:\Users\KARISHMA\Desktop\customer-churn-prediction\venv\Lib\site-packages\sklearn\base.py:1365: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Random Forest Accuracy:  0.54421651439407
              precision    recall  f1-score   support

           0       0.57      0.55      0.56      6168
           1       0.51      0.53      0.52      5434

    accuracy                           0.54     11602
   macro avg       0.54      0.54      0.54     11602
weighted avg       0.55      0.54      0.54     11602



what the above output means??
- with the help of random forest, we get 54.42% accuracy only. This shows the model isn't performing well.
- also the model is able to identify churners and non-churners by the same percentage

In [7]:
# XGBoost

xgb = XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42) #creating model
xgb.fit(X_train_bal, Y_train_bal) #training the model
y_pred_xgb = xgb.predict(X_test) #making predictions
print("XGBoost Accuracy: ", accuracy_score(Y_test, y_pred_xgb)) #evaluating model's performance
print(classification_report(Y_test, y_pred_xgb)) #metrics report

C:\Users\KARISHMA\Desktop\customer-churn-prediction\venv\Lib\site-packages\xgboost\training.py:199: UserWarning: [08:21:46] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:790: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBoost Accuracy:  0.543526978107223
              precision    recall  f1-score   support

           0       0.59      0.46      0.52      6168
           1       0.51      0.64      0.57      5434

    accuracy                           0.54     11602
   macro avg       0.55      0.55      0.54     11602
weighted avg       0.55      0.54      0.54     11602



In [9]:
#comparing results of various models

results = pd.DataFrame({
    'Model' : ['Logistic Regression', 'Random Forest', 'XGBoost'],
    'Accuracy' : [accuracy_score(Y_test, y_pred_lr),
        accuracy_score(Y_test, y_pred_rf),
        accuracy_score(Y_test, y_pred_xgb)]})

print(results)

                 Model  Accuracy
0  Logistic Regression  0.538786
1        Random Forest  0.544217
2              XGBoost  0.543527
